docs/design-documents/features/storage/TDBStore/TDBStore_design.md

```diff
@@ -0,0 +1,461 @@
+
+### Requirements and assumptions
+
+This feature requires a flash based block device such as `FlashIAPBlockDevice` or `SpifBlockDevice`. It can work on top of block devices that don't need erasing before writes, such as `HeapBlockDevice` or `SDBlockDevice`, but requires a flash simulator layer for this purpose, like the one offered by `FlashSimBlockDevice`. 
+
+# System architecture and high-level design
+
+## Design basics
+
+TDBStore includes the following design basics:
+- Sequential writes: All writes are made sequentially on the physical storage as records, superseding the previous ones for the same key.  
+- Memory layout - areas: The physical storage is divided equally into two areas - active and standby. All writes are made to the end of the active area's free space. When the active area is exhausted, a garbage collection process is invoked, copying only the up to date values of all keys to the standby area, and turning it active.
+- RAM table: Indexes all keys in RAM, thus allowing fast access to their records in the physical storage.   
+
+### Sequential writes
+All writes are made sequentially on the physical storage as records, superseding the previous ones for the same key. Each data record is written right after the last written one. If a key is updated, a new record with this key is written, overriding the previous value of this key. If a key is deleted, a new record with a "deleted" flag is added.
+Writes expect the storage to be erased. However, TDBStore takes the "erase as you go" approcah, meaning that when it crosses a sector boundary, it checks whether the next sector is erased, and if not - it gets erased. This saves a lot of time during initialization and garbage collection (see below). 
+
+### Memory layout and areas
+![TDBStore Areas](./TDBStore_areas.jpg)
+
+Each key is stored in a separate record on the active area. The first record in the area is the master record. Its main purpose is to hold an area version, protecting us against the case we have two valid areas (can happen in the extreme cases of power failures). 
+
+
+![TDBStore Record](./TDBStore_record.jpg)
+
+Record key and data are preceded by a 24-byte header. As the header is written separately, it needs to include a padding to the storage programming size.  
+Note: There are currently no external flags relevant to TDBStore.
+
+### Garbage collection
+Garbage collection (GC) is the process of compacting the records stored in the active area to the standby one, by copying only the most recent values of all the keys (without the ones marked as deleted). Then, the standby area becomes the active one and the previously active area is erased (not fully, only its first sector).  
+GC is invoked in the following cases:
+1. When the active area is exhausted.
+2. During initialization, when a corruption is found while scanning the active area. In this case, GC is performed up to the record preceding the corruption.
+
```

Comment:
GC is expensive and in an internal flash, it also blocking interrupts. Therefore I think it can be a good idea disabling GC when it founds out the area is full and there is nothing more to clean.
This will elimanate any further retries till a remove function is called.